### ISSUE TITLE
org.democratic-csi.nfs not found in the list of registered CSI drivers

### ISSUE THREAD

**Issue by RobbeDGreef** [democratic-csi/democratic-csi#270]
Hi, I have been trying to use my truenas as a persistent storage place for my talos os install. I have ran into a problem where kubernetes claims not to have the CSI driver I need registered, despite showing it when I run `kubctl get csidrivers` 

```
NAME                       ATTACHREQUIRED   PODINFOONMOUNT   STORAGECAPACITY   TOKENREQUESTS   REQUIRESREPUBLISH   MODES        AGE
org.democratic-csi.iscsi   false            true             false             <unset>         false               Persistent   175m
org.democratic-csi.nfs     false            true             false             <unset>         false               Persistent   3h3m
```

I have been trying with both nfs and iscsi but both give me the same result. For the sake of this issue I will focus on nfs since the issues seem closely related.

The issue seems very similar to #15 and #211, my kubelet path is `/var/lib/kubelet`.  The PVC is created and bound but when trying to mount it to a pod I get this error message when describing the pod:

    Warning  FailedMount  88s (x17 over 19m)   kubelet            MountVolume.MountDevice failed for volume "pvc-e3ad48dc-d722-43f6-9c1d-56f0a9705786" : kubernetes.io/csi: attacher.MountDevice failed to create newCsiDriverClient: driver name org.democratic-csi.nfs not found in the list of registered CSI drivers

My test pvc claim yaml looks like this:

```yaml
kind: PersistentVolumeClaim
apiVersion: v1
metadata:
  name: test-claim-nfs
  annotations:
    volume.beta.kubernetes.io/storage-class: "freenas-nfs-csi"
spec:
  storageClassName: freenas-nfs-csi
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
```

My test pod looks like this

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: hello-pod
spec:
  volumes:
  - name: local-storage
    persistentVolumeClaim:
      claimName: test-claim-nfs
  containers:
  - name: hello-container
    image: busybox
    command:
       - sh
       - -c
       - 'while true; do echo "`date` [`hostname`] Hello world." >> /mnt/store/greet.txt; sleep $(($RANDOM % 5 + 300)); done'
    volumeMounts:
    - mountPath: /mnt/store
      name: local-storage
```
My freenas-nfs.yaml:
```yaml
csiDriver:
  # should be globally unique for a given cluster
  name: "org.democratic-csi.nfs"

# add note here about volume expansion requirements
storageClasses:
- name: freenas-nfs-csi
  defaultClass: false
  reclaimPolicy: Delete
  volumeBindingMode: Immediate
  allowVolumeExpansion: true
  parameters:
    # for block-based storage can be ext3, ext4, xfs
    # for nfs should be nfs
    fsType: nfs
      
    # if true, volumes created from other snapshots will be
    # zfs send/received instead of zfs cloned
    # detachedVolumesFromSnapshots: "false"
    
    # if true, volumes created from other volumes will be
    # zfs send/received instead of zfs cloned
    # detachedVolumesFromVolumes: "false"

  mountOptions:
  - noatime
  - nfsvers=4
  secrets:
    provisioner-secret:
    controller-publish-secret:
    node-stage-secret:
    node-publish-secret:
    controller-expand-secret:

# if your cluster supports snapshots you may enable below
volumeSnapshotClasses: []
#- name: freenas-nfs-csi
#  parameters:
#  # if true, snapshots will be created with zfs send/receive
#  # detachedSnapshots: "false"
#  secrets:
#    snapshotter-secret:

driver:
  config:
    # please see the most up-to-date example of the corresponding config here:
    # https://github.com/democratic-csi/democratic-csi/tree/master/examples
    # YOU MUST COPY THE DATA HERE INLINE!
    driver: freenas-nfs
    instance_id:
    httpConnection:
      protocol: http
      host: server-ip
      port: 80
      # use only 1 of apiKey or username/password
      # if both are present, apiKey is preferred
      # apiKey is only available starting in TrueNAS-12
      #apiKey:
      username: root
      password: secret-password
      allowInsecure: true
      # use apiVersion 2 for TrueNAS-12 and up (will work on 11.x in some scenarios as well)
      # leave unset for auto-detection
      #apiVersion: 2
    sshConnection:
      host: server-ip
      port: 22
      username: root
      # use either password or key
      password: "secret-password"
      # privateKey: |
      #   -----BEGIN RSA PRIVATE KEY-----
      #   ...
      #   -----END RSA PRIVATE KEY-----
    zfs:
      # can be used to override defaults if necessary
      # the example below is useful for TrueNAS 12
      #cli:
      #  sudoEnabled: true
      #
      #  leave paths unset for auto-detection
      #  paths:
      #    zfs: /usr/local/sbin/zfs
      #    zpool: /usr/local/sbin/zpool
      #    sudo: /usr/local/bin/sudo
      #    chroot: /usr/sbin/chroot
      
      # can be used to set arbitrary values on the dataset/zvol
      # can use handlebars templates with the parameters from the storage class/CO
      #datasetProperties:
      #  "org.freenas:description": "{{ parameters.[csi.storage.k8s.io/pvc/namespace] }}/{{ parameters.[csi.storage.k8s.io/pvc/name] }}"
      #  "org.freenas:test": "{{ parameters.foo }}"
      #  "org.freenas:test2": "some value"

      datasetParentName: mainpool/k8s/nfs/vols
      # do NOT make datasetParentName and detachedSnapshotsDatasetParentName overlap
      # they may be siblings, but neither should be nested in the other
      detachedSnapshotsDatasetParentName: mainpool/k8s/nfs/snaps
      datasetEnableQuotas: true
      datasetEnableReservation: false
      datasetPermissionsMode: "0777"
      datasetPermissionsUser: 0
      datasetPermissionsGroup: 0
      #datasetPermissionsAcls:
      #- "-m everyone@:full_set:allow"
      #- "-m u:kube:full_set:allow"

    nfs:
      #shareCommentTemplate: "{{ parameters.[csi.storage.k8s.io/pvc/namespace] }}-{{ parameters.[csi.storage.k8s.io/pvc/name] }}"
      shareHost: server-ip
      shareAlldirs: false
      shareAllowedHosts: []
      shareAllowedNetworks: []
      shareMaprootUser: root
      shareMaprootGroup: wheel
      shareMapallUser: ""
      shareMapallGroup: ""
node:
  hostPID: true
  driver:
    extraEnv:
      - name: ISCSIADM_HOST_STRATEGY
        value: nsenter
      - name: ISCSIADM_HOST_PATH
        value: /usr/local/sbin/iscsiadm
    iscsiDirHostPath: /usr/local/etc/iscsi
    iscsiDirHostPathType: ""
```
I added the last node part since that is what the current README told me to do. I have been following [this](https://jonathangazeley.com/2021/01/05/using-truenas-to-provide-persistent-storage-for-kubernetes/) tutorial I found in the README.

When running `kubectl get pods -n democratic-csi` the pods seem healthy and do not show any errors when the logs are requested. I do not have any "nodes" running for some reason like other people (and the tutorial I am following) do. Only "controllers". This might be the source of the issue, but since I am very new to this I don't know.
```
NAME                                                   READY   STATUS    RESTARTS       AGE
zfs-iscsi-democratic-csi-controller-6db46dd8bc-r5m86   5/5     Running   15 (36m ago)   177m
zfs-nfs-democratic-csi-controller-5d78bbc94-ccbj8      5/5     Running   15 (36m ago)   81m
```

```
kubectl logs -n democratic-csi zfs-nfs-democratic-csi-controller-5d78bbc94-ccbj8:

Defaulted container "external-provisioner" out of: external-provisioner, external-resizer, external-snapshotter, csi-driver, csi-proxy
I0208 18:39:42.496242       1 feature_gate.go:245] feature gates: &{map[]}
I0208 18:39:42.496302       1 csi-provisioner.go:139] Version: v3.1.0
I0208 18:39:42.496308       1 csi-provisioner.go:162] Building kube configs for running in cluster...
I0208 18:39:42.500302       1 connection.go:154] Connecting to unix:///csi-data/csi.sock
I0208 18:39:46.177070       1 common.go:111] Probing CSI driver for readiness
I0208 18:39:46.177090       1 connection.go:183] GRPC call: /csi.v1.Identity/Probe
I0208 18:39:46.177095       1 connection.go:184] GRPC request: {}
I0208 18:39:46.187453       1 connection.go:186] GRPC response: {"ready":{"value":true}}
I0208 18:39:46.187603       1 connection.go:187] GRPC error: <nil>
I0208 18:39:46.187613       1 connection.go:183] GRPC call: /csi.v1.Identity/GetPluginInfo
I0208 18:39:46.187617       1 connection.go:184] GRPC request: {}
I0208 18:39:46.189401       1 connection.go:186] GRPC response: {"name":"org.democratic-csi.nfs","vendor_version":"1.7.7"}
I0208 18:39:46.189463       1 connection.go:187] GRPC error: <nil>
I0208 18:39:46.189475       1 csi-provisioner.go:206] Detected CSI driver org.democratic-csi.nfs
I0208 18:39:46.189485       1 connection.go:183] GRPC call: /csi.v1.Identity/GetPluginCapabilities
I0208 18:39:46.189491       1 connection.go:184] GRPC request: {}
I0208 18:39:46.191083       1 connection.go:186] GRPC response: {"capabilities":[{"Type":{"Service":{"type":1}}},{"Type":{"VolumeExpansion":{"type":1}}}]}
I0208 18:39:46.191203       1 connection.go:187] GRPC error: <nil>
I0208 18:39:46.191211       1 connection.go:183] GRPC call: /csi.v1.Controller/ControllerGetCapabilities
I0208 18:39:46.191214       1 connection.go:184] GRPC request: {}
I0208 18:39:46.192863       1 connection.go:186] GRPC response: {"capabilities":[{"Type":{"Rpc":{"type":1}}},{"Type":{"Rpc":{"type":3}}},{"Type":{"Rpc":{"type":4}}},{"Type":{"Rpc":{"type":5}}},{"Type":{"Rpc":{"type":6}}},{"Type":{"Rpc":{"type":7}}},{"Type":{"Rpc":{"type":9}}},{"Type":{"Rpc":{"type":12}}},{"Type":{"Rpc":{"type":13}}}]}
I0208 18:39:46.192984       1 connection.go:187] GRPC error: <nil>
I0208 18:39:46.193619       1 csi-provisioner.go:278] CSI driver does not support PUBLISH_UNPUBLISH_VOLUME, not watching VolumeAttachments
I0208 18:39:46.194292       1 controller.go:732] Using saving PVs to API server in background
I0208 18:39:46.194647       1 leaderelection.go:248] attempting to acquire leader lease democratic-csi/org-democratic-csi-nfs...
I0208 18:39:46.277413       1 leader_election.go:212] new leader detected, current leader: zfs-nfs-democratic-csi-controller-5d78bbc94-ccbj8
I0208 18:39:46.277427       1 leaderelection.go:258] successfully acquired lease democratic-csi/org-democratic-csi-nfs
I0208 18:39:46.277516       1 leader_election.go:205] became leader, starting
I0208 18:39:46.278239       1 reflector.go:219] Starting reflector *v1.PersistentVolumeClaim (15m0s) from k8s.io/client-go/informers/factory.go:134
I0208 18:39:46.278249       1 reflector.go:219] Starting reflector *v1.StorageClass (1h0m0s) from k8s.io/client-go/informers/factory.go:134
I0208 18:39:46.278273       1 reflector.go:255] Listing and watching *v1.StorageClass from k8s.io/client-go/informers/factory.go:134
I0208 18:39:46.278273       1 reflector.go:255] Listing and watching *v1.PersistentVolumeClaim from k8s.io/client-go/informers/factory.go:134
I0208 18:39:46.328247       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 18:39:46.378316       1 shared_informer.go:270] caches populated
I0208 18:39:46.378331       1 shared_informer.go:270] caches populated
I0208 18:39:46.378340       1 controller.go:811] Starting provisioner controller org.democratic-csi.nfs_zfs-nfs-democratic-csi-controller-5d78bbc94-ccbj8_9230784d-40d5-4e09-913a-1d0264ae9109!
I0208 18:39:46.378375       1 clone_controller.go:66] Starting CloningProtection controller
I0208 18:39:46.378398       1 clone_controller.go:82] Started CloningProtection controller
I0208 18:39:46.378450       1 volume_store.go:97] Starting save volume queue
I0208 18:39:46.378547       1 reflector.go:219] Starting reflector *v1.PersistentVolume (15m0s) from sigs.k8s.io/sig-storage-lib-external-provisioner/v8/controller/controller.go:845
I0208 18:39:46.378556       1 reflector.go:219] Starting reflector *v1.StorageClass (15m0s) from sigs.k8s.io/sig-storage-lib-external-provisioner/v8/controller/controller.go:848
I0208 18:39:46.378570       1 reflector.go:255] Listing and watching *v1.StorageClass from sigs.k8s.io/sig-storage-lib-external-provisioner/v8/controller/controller.go:848
I0208 18:39:46.378560       1 reflector.go:255] Listing and watching *v1.PersistentVolume from sigs.k8s.io/sig-storage-lib-external-provisioner/v8/controller/controller.go:845
I0208 18:39:46.479181       1 shared_informer.go:270] caches populated
I0208 18:39:46.479261       1 controller.go:860] Started provisioner controller org.democratic-csi.nfs_zfs-nfs-democratic-csi-controller-5d78bbc94-ccbj8_9230784d-40d5-4e09-913a-1d0264ae9109!
I0208 18:39:51.340414       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
...
I0208 18:46:50.884997       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 18:46:53.411665       1 reflector.go:536] sigs.k8s.io/sig-storage-lib-external-provisioner/v8/controller/controller.go:848: Watch close - *v1.StorageClass total 8 items received
I0208 18:46:55.960699       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 18:46:57.311232       1 reflector.go:536] k8s.io/client-go/informers/factory.go:134: Watch close - *v1.PersistentVolumeClaim total 9 items received
I0208 18:47:01.139817       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
...
I0208 18:48:02.539043       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 18:48:05.312418       1 reflector.go:536] k8s.io/client-go/informers/factory.go:134: Watch close - *v1.StorageClass total 9 items received
I0208 18:48:07.839024       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 18:48:12.412015       1 reflector.go:536] sigs.k8s.io/sig-storage-lib-external-provisioner/v8/controller/controller.go:845: Watch close - *v1.PersistentVolume total 10 items received
I0208 18:48:12.884606       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
...
I0208 18:52:12.329492       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 18:52:12.414825       1 reflector.go:536] sigs.k8s.io/sig-storage-lib-external-provisioner/v8/controller/controller.go:848: Watch close - *v1.StorageClass total 7 items received
I0208 18:52:17.358762       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 18:52:22.383410       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 18:52:27.409613       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 18:52:32.437706       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 18:52:37.479718       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 18:52:42.502265       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 18:52:43.315502       1 reflector.go:536] k8s.io/client-go/informers/factory.go:134: Watch close - *v1.PersistentVolumeClaim total 7 items received
I0208 18:52:47.529610       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
...
I0208 18:53:32.758156       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 18:53:34.316358       1 reflector.go:536] k8s.io/client-go/informers/factory.go:134: Watch close - *v1.StorageClass total 7 items received
I0208 18:53:37.787632       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
...
I0208 18:54:38.073746       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 18:54:42.417133       1 reflector.go:536] sigs.k8s.io/sig-storage-lib-external-provisioner/v8/controller/controller.go:845: Watch close - *v1.PersistentVolume total 8 items received
I0208 18:54:43.085937       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 18:54:46.280600       1 reflector.go:382] k8s.io/client-go/informers/factory.go:134: forcing resync
I0208 18:54:46.380898       1 reflector.go:382] sigs.k8s.io/sig-storage-lib-external-provisioner/v8/controller/controller.go:845: forcing resync
I0208 18:54:48.097442       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
...
I0208 18:59:34.001495       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 18:59:38.321663       1 reflector.go:536] k8s.io/client-go/informers/factory.go:134: Watch close - *v1.StorageClass total 8 items received
I0208 18:59:39.024738       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 18:59:44.040016       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 18:59:46.420941       1 reflector.go:536] sigs.k8s.io/sig-storage-lib-external-provisioner/v8/controller/controller.go:848: Watch close - *v1.StorageClass total 9 items received
I0208 18:59:49.176441       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
...
I0208 19:01:31.556440       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 19:01:36.421711       1 reflector.go:536] sigs.k8s.io/sig-storage-lib-external-provisioner/v8/controller/controller.go:845: Watch close - *v1.PersistentVolume total 8 items received
I0208 19:01:36.811302       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 19:01:41.824189       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 19:01:46.874662       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 19:01:47.321613       1 reflector.go:536] k8s.io/client-go/informers/factory.go:134: Watch close - *v1.PersistentVolumeClaim total 11 items received
I0208 19:01:51.966939       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
...
I0208 19:06:13.013455       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 19:06:13.327216       1 reflector.go:536] k8s.io/client-go/informers/factory.go:134: Watch close - *v1.StorageClass total 7 items received
I0208 19:06:18.088610       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
...
I0208 19:07:03.274107       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 19:07:06.427376       1 reflector.go:536] sigs.k8s.io/sig-storage-lib-external-provisioner/v8/controller/controller.go:848: Watch close - *v1.StorageClass total 9 items received
I0208 19:07:08.300526       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
...
I0208 19:07:58.574239       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 19:08:00.427692       1 reflector.go:536] sigs.k8s.io/sig-storage-lib-external-provisioner/v8/controller/controller.go:845: Watch close - *v1.PersistentVolume total 8 items received
I0208 19:08:03.595147       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 19:08:08.621157       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 19:08:13.649314       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 19:08:14.327909       1 reflector.go:536] k8s.io/client-go/informers/factory.go:134: Watch close - *v1.PersistentVolumeClaim total 8 items received
I0208 19:08:18.678969       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
...
I0208 19:09:44.090466       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 19:09:46.281237       1 reflector.go:382] k8s.io/client-go/informers/factory.go:134: forcing resync
I0208 19:09:46.381462       1 reflector.go:382] sigs.k8s.io/sig-storage-lib-external-provisioner/v8/controller/controller.go:845: forcing resync
I0208 19:09:49.111116       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
...
I0208 19:13:07.537261       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 19:13:11.434464       1 reflector.go:536] sigs.k8s.io/sig-storage-lib-external-provisioner/v8/controller/controller.go:848: Watch close - *v1.StorageClass total 8 items received
I0208 19:13:12.566088       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
...
I0208 19:13:57.767856       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 19:14:00.434475       1 reflector.go:536] sigs.k8s.io/sig-storage-lib-external-provisioner/v8/controller/controller.go:845: Watch close - *v1.PersistentVolume total 7 items received
I0208 19:14:02.801834       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
...
I0208 19:14:32.978615       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 19:14:36.335015       1 reflector.go:536] k8s.io/client-go/informers/factory.go:134: Watch close - *v1.StorageClass total 10 items received
I0208 19:14:38.005498       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
...
I0208 19:14:58.109029       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
I0208 19:15:02.333413       1 reflector.go:536] k8s.io/client-go/informers/factory.go:134: Watch close - *v1.PersistentVolumeClaim total 8 items received
I0208 19:15:03.142077       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
...
I0208 19:17:18.851576       1 leaderelection.go:278] successfully renewed lease democratic-csi/org-democratic-csi-nfs
```
(note that I cut out some repeating lines to shorten the log)

**Comment by travisghansen** [comment]
Do you know what path talos uses for kubelet?

- https://github.com/democratic-csi/democratic-csi#talos
- https://github.com/democratic-csi/democratic-csi#a-note-on-non-standard-kubelet-paths

**Comment by RobbeDGreef** [comment]
`/var/lib/kubelet` which is considered the standard path to my knowing.

**Comment by travisghansen** [comment]
Can you send the list of all pods running in the ns?

**Comment by RobbeDGreef** [comment]
The democratic-csi namespace is running these pods 

```
NAME                                                   READY   STATUS    RESTARTS       AGE
zfs-iscsi-democratic-csi-controller-6db46dd8bc-r5m86   5/5     Running   15 (36m ago)   177m
zfs-nfs-democratic-csi-controller-5d78bbc94-ccbj8      5/5     Running   15 (36m ago)   81m
```

Like I said in the description I don't fully understand why I am only running controllers and no nodes. It might have something to do with it but I couldn't find any info about it.

**Comment by RobbeDGreef** [comment]
kubectl get daemonsets -n democratic-csi

```
NAME                            DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE
zfs-iscsi-democratic-csi-node   1         0         0       0            0           kubernetes.io/os=linux   3h45m
zfs-nfs-democratic-csi-node     1         0         0       0            0           kubernetes.io/os=linux   3h53m
```

```
kubectl describe daemonset zfs-nfs-democratic-csi-node -n democratic-csi
Name:           zfs-nfs-democratic-csi-node
Selector:       app.kubernetes.io/component=node-linux,app.kubernetes.io/csi-role=node,app.kubernetes.io/instance=zfs-nfs,app.kubernetes.io/managed-by=Helm,app.kubernetes.io/name=democratic-csi
Node-Selector:  kubernetes.io/os=linux
Labels:         app.kubernetes.io/component=node-linux
                app.kubernetes.io/csi-role=node
                app.kubernetes.io/instance=zfs-nfs
                app.kubernetes.io/managed-by=Helm
                app.kubernetes.io/name=democratic-csi
                helm.sh/chart=democratic-csi-0.13.5
Annotations:    deprecated.daemonset.template.generation: 4
                meta.helm.sh/release-name: zfs-nfs
                meta.helm.sh/release-namespace: democratic-csi
Desired Number of Nodes Scheduled: 1
Current Number of Nodes Scheduled: 0
Number of Nodes Scheduled with Up-to-date Pods: 0
Number of Nodes Scheduled with Available Pods: 0
Number of Nodes Misscheduled: 0
Pods Status:  0 Running / 0 Waiting / 0 Succeeded / 0 Failed
Pod Template:
  Labels:           app.kubernetes.io/component=node-linux
                    app.kubernetes.io/csi-role=node
                    app.kubernetes.io/instance=zfs-nfs
                    app.kubernetes.io/managed-by=Helm
                    app.kubernetes.io/name=democratic-csi
  Annotations:      checksum/configmap: 4c738d703c418eb6a75fa3a097249ef9bf02c0678a385c5bb41bd6c2416beef9
                    checksum/secret: 99e60fa053b1dda6f65868666b992f8e6274f2d0e6553e445082ba6cd90cafd1
  Service Account:  zfs-nfs-democratic-csi-node-sa
  Containers:
   csi-driver:
    Image:      docker.io/democraticcsi/democratic-csi:latest
    Port:       <none>
    Host Port:  <none>
    Args:
      --csi-version=1.5.0
      --csi-name=org.democratic-csi.nfs
      --driver-config-file=/config/driver-config-file.yaml
      --log-level=info
      --csi-mode=node
      --server-socket=/csi-data/csi.sock.internal
    Liveness:  exec [bin/liveness-probe --csi-version=1.5.0 --csi-address=/csi-data/csi.sock.internal] delay=10s timeout=15s period=60s #success=1 #failure=3
    Environment:
      CSI_NODE_ID:              (v1:spec.nodeName)
      NODE_EXTRA_CA_CERTS:     /tmp/certs/extra-ca-certs.crt
      ISCSIADM_HOST_STRATEGY:  nsenter
      ISCSIADM_HOST_PATH:      /usr/local/sbin/iscsiadm
    Mounts:
      /config from config (rw)
      /csi-data from socket-dir (rw)
      /dev from dev-dir (rw)
      /etc/localtime from localtime (ro)
      /host from host-dir (rw)
      /lib/modules from modules-dir (ro)
      /run/udev from udev-data (rw)
      /sys from sys-dir (rw)
      /tmp/certs from extra-ca-certs (rw)
      /var/lib/kubelet from kubelet-dir (rw)
   csi-proxy:
    Image:      docker.io/democraticcsi/csi-grpc-proxy:v0.5.3
    Port:       <none>
    Host Port:  <none>
    Environment:
      BIND_TO:   unix:///csi-data/csi.sock
      PROXY_TO:  unix:///csi-data/csi.sock.internal
    Mounts:
      /csi-data from socket-dir (rw)
   driver-registrar:
    Image:      k8s.gcr.io/sig-storage/csi-node-driver-registrar:v2.5.1
    Port:       <none>
    Host Port:  <none>
    Args:
      --v=5
      --csi-address=/csi-data/csi.sock
      --kubelet-registration-path=/var/lib/kubelet/plugins/org.democratic-csi.nfs/csi.sock
    Liveness:  exec [/csi-node-driver-registrar --kubelet-registration-path=/var/lib/kubelet/plugins/org.democratic-csi.nfs/csi.sock --mode=kubelet-registration-probe] delay=0s timeout=1s period=10s #success=1 #failure=3
    Environment:
      KUBE_NODE_NAME:   (v1:spec.nodeName)
    Mounts:
      /csi-data from socket-dir (rw)
      /registration from registration-dir (rw)
      /var/lib/kubelet from kubelet-dir (rw)
   cleanup:
    Image:      docker.io/busybox:1.32.0
    Port:       <none>
    Host Port:  <none>
    Command:
      /bin/sh
      -c
      --
    Args:
      while true; do sleep 2; done;
    Environment:  <none>
    Mounts:
      /plugins from plugins-dir (rw)
      /registration from registration-dir (rw)
  Volumes:
   socket-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/kubelet/plugins/org.democratic-csi.nfs
    HostPathType:  DirectoryOrCreate
   plugins-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/kubelet/plugins
    HostPathType:  Directory
   registration-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/kubelet/plugins_registry
    HostPathType:  Directory
   kubelet-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /var/lib/kubelet
    HostPathType:  Directory
   dev-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /dev
    HostPathType:  Directory
   modules-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /lib/modules
    HostPathType:  
   localtime:
    Type:          HostPath (bare host directory volume)
    Path:          /etc/localtime
    HostPathType:  
   udev-data:
    Type:          HostPath (bare host directory volume)
    Path:          /run/udev
    HostPathType:  
   sys-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /sys
    HostPathType:  Directory
   host-dir:
    Type:          HostPath (bare host directory volume)
    Path:          /
    HostPathType:  Directory
   config:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  zfs-nfs-democratic-csi-driver-config
    Optional:    false
   extra-ca-certs:
    Type:      ConfigMap (a volume populated by a ConfigMap)
    Name:      zfs-nfs-democratic-csi
    Optional:  false
Events:
  Type     Reason        Age   From                  Message
  ----     ------        ----  ----                  -------
  Warning  FailedCreate  47m   daemonset-controller  Error creating: pods "zfs-nfs-democratic-csi-node-z7vqd" is forbidden: violates PodSecurity "baseline:latest": non-default capabilities (container "csi-driver" must not include "SYS_ADMIN" in securityContext.capabilities.add), host namespaces (hostNetwork=true, hostPID=true, hostIPC=true), hostPath volumes (volumes "socket-dir", "plugins-dir", "registration-dir", "kubelet-dir", "dev-dir", "modules-dir", "localtime", "udev-data", "sys-dir", "host-dir"), privileged (container "csi-driver" must not set securityContext.privileged=true)
  Warning  FailedCreate  30m   daemonset-controller  Error creating: pods "zfs-nfs-democratic-csi-node-9sdpz" is forbidden: violates PodSecurity "baseline:latest": non-default capabilities (container "csi-driver" must not include "SYS_ADMIN" in securityContext.capabilities.add), host namespaces (hostNetwork=true, hostPID=true, hostIPC=true), hostPath volumes (volumes "socket-dir", "plugins-dir", "registration-dir", "kubelet-dir", "dev-dir", "modules-dir", "localtime", "udev-data", "sys-dir", "host-dir"), privileged (container "csi-driver" must not set securityContext.privileged=true)
  Warning  FailedCreate  14m   daemonset-controller  Error creating: pods "zfs-nfs-democratic-csi-node-ccsgp" is forbidden: violates PodSecurity "baseline:latest": non-default capabilities (container "csi-driver" must not include "SYS_ADMIN" in securityContext.capabilities.add), host namespaces (hostNetwork=true, hostPID=true, hostIPC=true), hostPath volumes (volumes "socket-dir", "plugins-dir", "registration-dir", "kubelet-dir", "dev-dir", "modules-dir", "localtime", "udev-data", "sys-dir", "host-dir"), privileged (container "csi-driver" must not set securityContext.privileged=true)
```

These security warnings seem to be the source of the issue, I don't immediately know how to resolve them but I will look into it.

**Comment by RobbeDGreef** [comment]
It is working now, thanks a lot!

I simply labeled the democratic-csi namespace and enforced privileged behavior according to [this link](https://kubernetes.io/docs/tasks/configure-pod-container/enforce-standards-namespace-labels/).

After that and a reboot the nodes were created and everything was working.

**Comment by travisghansen** [comment]
Great! Can you send over specific details so I can a note to the README?

**Comment by RobbeDGreef** [comment]
Sure so the root cause of the issue was that the democratic csi nodes weren't showing up and that was due to kubernetes pod security not allowing the pods to be created. This can be seen in the logs from the daemonset above. 
The solution I went with is to label the democratic csi namespace to have less harsh security rules with

`kubectl label --overwrite namespace democratic-csi pod-security.kubernetes.io/enforce=privileged`

Since it is only for that specific namespace that contains democratic csi stuff, I think it should be fine security wise. If not then let me know.

After that I just rebooted the talos machine and everything was working fine, nodes showed up and the CSI drivers were found by kubernetes.
